compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUSplitReduction.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-split-reduction"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+class LLVMCPUSplitReductionPass
+    : public LLVMCPUSplitReductionBase<LLVMCPUSplitReductionPass> {
+ public:
+  LLVMCPUSplitReductionPass(bool fpReductionReordering)
+      : fpReductionReordering(fpReductionReordering) {}
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, scf::SCFDialect>();
+  }
+  void runOnOperation() override;
+
+ private:
+  bool fpReductionReordering = false;
+};
+
+void LLVMCPUSplitReductionPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  SmallVector<Operation *> computeOps = getComputeOps(funcOp);
+  FailureOr<IREE::Codegen::LoweringConfigAttr> maybeLoweringConfig =
+      getLoweringConfig(computeOps);
+  if (failed(maybeLoweringConfig)) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "can't find lowering_config, skip SplitReduction");
+    return;
+  }
+  auto reductionSizes = maybeLoweringConfig.value().getTileSizeVals().back();
+  if (reductionSizes.empty()) {
+    LLVM_DEBUG(
+        llvm::dbgs()
+        << "the list of reduction tiling sizes is empty, skip SplitReduction");
+    return;
+  }
+  int64_t reductionRatio = reductionSizes.back();
+
+  SmallVector<linalg::GenericOp> candidates;
+  funcOp.walk([&](linalg::GenericOp op) { candidates.push_back(op); });
+  for (auto genericOp : candidates) {
```

Comment:
It might be better to move the handling of each candidate to a separate function that accepts the RewriterBase as argument. You can make that method return LogicalResult and use notifyMatchFailure. There should a way to dump out the failure messages from the rewriter in the caller when the function returns failure